### DISCUSSION TITLE
How can I filter deep?

### DISCUSSION THREAD

**Discussion by ilterugur** [TriPSs/nestjs-query#113] [Q&A]
Hi, I have a schema which stores some nested information:

**flair.schema.ts:**
```
import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
import { BaseSchema } from '../../mongoose';
import { Schema as MongooseSchema } from 'mongoose';
import mongooseSlugPlugin from 'mongoose-slug-plugin';
import {
  CreateFlairTranslationsInputDTO,
  CreateFlairPaletteInputDTO,
} from '../dto';

@Schema({
  timestamps: true,
  collection: 'flairs',
})
export class Flair extends BaseSchema<Flair> {
  _id: MongooseSchema.Types.ObjectId;

  slug: string;

  @Prop({
    required: true,
    type: CreateFlairPaletteInputDTO,
  })
  palette: CreateFlairPaletteInputDTO;

  @Prop({
    required: true,
    type: CreateFlairTranslationsInputDTO,
  })
  translations: CreateFlairTranslationsInputDTO;

  @Prop()
  createdAt: Date;

  @Prop()
  updatedAt: Date;
}

export const FlairSchema = SchemaFactory.createForClass(Flair);
FlairSchema.plugin(mongooseSlugPlugin, {
  tmpl: '<%=translations.en.name%>',
  slugOptions: { "'": '' },
});
FlairSchema.virtual('catalogItems', {
  ref: 'CatalogItem',
  localField: '_id',
  foreignField: 'flairs',
});
```

**flair.dto.ts:**
```
import { ID, ObjectType } from '@nestjs/graphql';
import { Schema as MongooseSchema } from 'mongoose';
import {
  CursorConnection,
  FilterableField,
  IDField,
  ObjectId,
  QueryOptions,
} from '@ptc-org/nestjs-query-graphql';
import { CatalogItemDTO } from '../../catalog/dto';
import { UpdateFlairTranslationsInputDTO } from './translations/update-flair-translations.input.dto';
import { UpdateFlairPaletteInputDTO } from './update-flair-palette-input.dto';

@CursorConnection('catalogItems', () => CatalogItemDTO, {
  disableUpdate: true,
  disableRemove: true,
})
@ObjectType('Flair')
@QueryOptions({ enableTotalCount: true })
export class FlairDTO {
  @IDField(() => ID)
  @ObjectId()
  _id: MongooseSchema.Types.ObjectId;

  @FilterableField(() => String, { description: 'Flair slug' })
  slug: string;

  @FilterableField(() => UpdateFlairPaletteInputDTO, {
    description: 'Flair palette',
  })
  palette: UpdateFlairPaletteInputDTO;

  @FilterableField(() => UpdateFlairTranslationsInputDTO, {
    description: 'Flair translations',
  })
  translations: UpdateFlairTranslationsInputDTO;
}
```

**update-flair-palette.input.dto.ts:**
```
import { InputType, ObjectType } from '@nestjs/graphql';
import { FilterableField } from '@ptc-org/nestjs-query-graphql';

@InputType('UpdateFlairPaletteInput')
@ObjectType('FlairPalette')
export class UpdateFlairPaletteInputDTO {
  @FilterableField({ nullable: true })
  background: string;

  @FilterableField({ nullable: true })
  text: string;
}
```

I want to use filter like: flairs.filter.palette.background eq e6ab09 but when I try to do that it returns an empty array and it queries something like:
`Mongoose: flairs.find({ '$and': [ { palette: { '$regex': /\[object Object\]/ } } ]}, { limit: 11, skip: 0 })`
(I've found this query using mongoose debug)

What should I do?

**Comment by TriPSs** [comment]
I'm not quite know with Mongoose but is `palette` a relation of in the stored object?

**Reply by ilterugur** [reply]
> I'm not quite know with Mongoose but is `palette` a relation of in the stored object?

Sorry, I realized that I forgot to add my schema to the question and now I've edited the question and added my mongoose schema

palette is not a relation in my MongoDB database. It's just the name of a key which stores an object as the value.

Example:
```
{
  "_id": {
    "$oid": "640a61688c1d3b0c1884abb1"
  },
  "palette": {
    "background": "fcba03",
    "text": "ffffff"
  },
  "translations": {
    "en": {
      "description": "Flair test description",
      "name": "Flair test name"
    }
  },
  "slug_history": [
    "flair-test-name"
  ],
  "slug": "flair-test-name",
  "createdAt": {
    "$date": {
      "$numberLong": "1678401896547"
    }
  },
  "updatedAt": {
    "$date": {
      "$numberLong": "1678401896547"
    }
  },
  "__v": 0
}
```

In this example, I want to filter flairs with "fcba03" background which is stored on palette.background. It's nested and I want to query based on "background" key but it queries on "palette" key which is "background" key's parent.

My MongoDB log:
`ctx":"conn472","msg":"Only one plan is available","attr":{"query":"ns: testdb_nestjs.flairs query: { $and: [ { palette: { $eq: { background: \"fcba03\" } } } ] } sort: {} projection: {} limit: 11","planSummary":"COLLSCAN"`

It shouldn't be like this:
`testdb_nestjs.flairs query: { $and: [ { palette: { $eq: { background: \"fcba03\" } } } ] `

But instead, it should look something like this:
`testdb_nestjs.flairs query: { $and: [ { palette: { background: { $eq:  \"fcba03\" } } } ] `

**Reply by ilterugur** [reply]
My graphql query:
```
query MyQuery {
  flairs(filter: {palette: {eq: {background: "fcba03"}}}) {
    edges {
      cursor
      node {
        slug
        _id
        translations {
          en {
            description
            name
          }
        }
      }
    }
  }
}
```

You see that filter is:
`(filter: {palette: {eq: {background: "fcba03"}}})`

But it should be:
`(filter: {palette: {background: {eq: "fcba03"}}})`

I want to filter on background but I don't see option for "background" under "filter" on _graphiql_ interface. It shows operators like "eq", "like", "ilike" but it should also show nested object's attributes if I marked them with "FilterableField" decorator
